### PR TITLE
remove queue bindings which are no longer part of the queue declaration

### PR DESCRIPTION
**Pull request by skaes**
ATTENTION: this assumes that queue declarations are not shared between different applications.

Note that the implementation is only eventually consistent in that
during an application deployment it is possible that old clients can cause
deleted bindings to reappear due to queue declarations. As a consequence,
consumers need to be able to handle messages messages routed previously,
until all clients have been restarted.

### CODE REVIEW

File: lib/beetle/dead_lettering.rb [hunk 1]
```diff
@@ -68,6 +72,58 @@ def set_queue_policy!(server, queue_name, options={})
...
+
+      response = run_rabbit_http_request(request_url, request) do |http|
+        http.request(request)
+      end
+
+      unless %w(200 201 204).include?(response.code)
+        log_error("Failed to remove obsolete binding for queue #{queue_name}", response)
+        raise FailedRabbitRequest.new("Could not retrieve queue bindings")
+      end
+    end
```
Comment (boosty) [inline lib/beetle/dead_lettering.rb]: As discussed, I think the `DeadLettering` class is now doing too much. We should find a new home for all queue management related methods.

Comment (skaes) [reply lib/beetle/dead_lettering.rb]: renamed to QueueProperties

File: lib/beetle/client.rb [hunk 2]
```diff
@@ -87,16 +87,16 @@ def register_exchange(name, options={})
     #   the binding key (defaults to the name of the queue)
     # [<tt>:lazy</tt>]
     #   whether the queue should use lazy mode (defaults to <tt>config.lazy_queues_enabled</tt>)
-    # [<tt>:dead_lettering</tt>]
-    #   whether the queue should use dead lettering (defaults to <tt>config.dead_lettering_enabled</tt>)
+    # [<tt>:queue_properties</tt>]
+    #   whether the queue should use dead lettering (defaults to <tt>config.queue_properties_enabled</tt>)
```
Comment (boosty) [inline lib/beetle/client.rb]: The option name and description don't fit together.

File: lib/beetle/client.rb [hunk 3]
```diff
@@ -55,15 +55,15 @@ def initialize(config = Beetle.config)
...
+      @queue_properties = QueueProperties.new(config)
       load_brokers_from_config
       register_exchange(config.beetle_policy_exchange_name)
       # make sure dead lettering is false for the policy update queue
       register_queue(
         config.beetle_policy_updates_queue_name,
         :exchange => config.beetle_policy_exchange_name,
         :key => config.beetle_policy_updates_routing_key,
-        :dead_lettering => false,
+        :queue_properties => false,
```
Comment (boosty) [inline lib/beetle/client.rb]: The comment above still says *make sure dead lettering is false for the policy update queue*, but *dead lettering* does not appear anymore.

To be honest, this interface is not very intuitive. `:queue_properties` sounds like an options hash to me, not like boolean as used here. And then why have an extra `:lazy` option. Shouldn't this be part of `:queue_properties`?

File: DEAD_LETTERING.md [hunk 4]
```diff
@@ -45,7 +45,7 @@ example:
 client = Beetle::Client.new.configure do |config|
   config.handler(Beetle.config.beetle_policy_updates_queue_name) do |message|
     Beetle.config.logger.info "received beetle policy update message': #{message.data}"
-    client.set_queue_policies!(JSON.parse(message.data))
+    client.update_queue_properties!(JSON.parse(message.data))
```
[the remark this answers is not in the document]
Comment (skaes) [reply DEAD_LETTERING.md]: fixed

[the remark this answers is not in the document]
Comment (skaes) [reply DEAD_LETTERING.md]: fixed

File: lib/beetle/client.rb (hunk 2 above)
Comment (skaes) [reply lib/beetle/client.rb]: fixed

File: lib/beetle/client.rb (hunk 3 above)
Comment (skaes) [reply lib/beetle/client.rb]: renamed too many instances of dead_lettering with search and replace. fixed.

File: test/beetle/subscriber_test.rb [hunk 5]
```diff
@@ -225,7 +225,7 @@ def setup
   end
 
 
-  class DeadLetteringCallBackExecutionTest < Minitest::Test
+  class QueuePropertiesCallBackExecutionTest < Minitest::Test
```
Comment (boosty) [inline test/beetle/subscriber_test.rb]: The test class has little to do with the `QueueProperties`, so I think we could just keep the old name.

File: lib/beetle/queue_properties.rb [hunk 6]
```diff
@@ -11,8 +11,12 @@ def initialize(config)
...
     end
 
-    def set_queue_policies!(options)
-      # logger.debug "Setting queue policies: #{options.inspect}"
+    def vhost
+      CGI.escape(@config.vhost)
+    end
+
+    def update_queue_properties!(options)
+      # logger.debug "Updating queue properties: #{options.inspect}"
```
[the remark this answers is not in the document]
Comment (skaes) [reply lib/beetle/queue_properties.rb]: enabled

File: test/beetle/client_test.rb [hunk 7]
```diff
@@ -356,10 +356,10 @@ class ClientTest < Minitest::Test
       client.stop_listening
     end
 
-    test "should delegate set_queue_policies! to the dead lettering instance" do
+    test "should delegate update_queue_properties! to the dead lettering instance" do
```
[the remark this answers is not in the document]
Comment (skaes) [reply test/beetle/client_test.rb]: fixed

File: test/beetle/subscriber_test.rb (hunk 5 above)
Comment (skaes) [reply test/beetle/subscriber_test.rb]: changed

**Pull request merged**